include/gridtools/storage/adapter/python_sid_adapter.hpp

```diff
@@ -0,0 +1,83 @@
+/*
+ * GridTools
+ *
+ * Copyright (c) 2014-2019, ETH Zurich
+ * All rights reserved.
+ *
+ * Please, refer to the LICENSE file in the root directory.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+#pragma once
+
+#include <array>
+#include <cassert>
+#include <cstdlib>
+#include <stdexcept>
+#include <string>
+#include <type_traits>
+#include <utility>
+
+#include <pybind11/pybind11.h>
+
+#include "../../common/integral_constant.hpp"
+#include "../../sid/simple_ptr_holder.hpp"
+
+namespace gridtools {
+    namespace python_sid_adapter_impl_ {
+        template <size_t, class>
+        struct kind {};
+
+        template <class T, size_t Dim, class Kind>
+        struct wrapper {
+            pybind11::buffer_info m_info;
+
+            friend sid::simple_ptr_holder<T *> sid_get_origin(wrapper const &obj) {
+                return {reinterpret_cast<T *>(obj.m_info.ptr)};
+            }
+            friend std::array<pybind11::ssize_t, Dim> sid_get_strides(wrapper const &obj) {
+                std::array<pybind11::ssize_t, Dim> res;
+                assert(obj.m_info.strides.size() == Dim);
+                for (std::size_t i = 0; i != Dim; ++i) {
+                    assert(obj.m_info.strides[i] % obj.m_info.itemsize == 0);
+                    res[i] = obj.m_info.strides[i] / obj.m_info.itemsize;
+                }
+                return res;
+            }
+            friend std::array<integral_constant<pybind11::ssize_t, 0>, Dim> sid_get_lower_bounds(wrapper const &) {
```

Comment:
@egparedes and I discussed the API for the generated code again in the morning. We talked about how the domain is passed to generated code.

- In gt4py this is 3 numbers, the sizes of the domain. To distinguish compute domain vs halo, the origin is shifted in the storages.
- I think long ago (with @lukasm91) we came to the conclusion that this makes sense. We'd go for this approach for the generated code.

Can we then use sid lower and upper bounds for bounds checking out of bounds?

Reply:
You definitely can. However there are nuances here: in general case `get_lower_bounds`/`get_upper_bounds` return  the `hymap`s that are not necessarily contain all the iteratable dimensions of the data. The absence of the dimension means that there is no bounds along it. For example `global_parameter` sid returns an empty tuple for both lower and upper bounds. 

See here: https://github.com/GridTools/gridtools/blob/master/include/gridtools/stencil/frontend/run.hpp#L218 for an example of an accurate bounds checking.

Reply:
If it could be related to the question: if you need to shift the data that comes from python by halo, you can chain `as_sid` with the `sid::shift_sid_origin`.

Reply:
> See here: https://github.com/GridTools/gridtools/blob/master/include/gridtools/stencil/frontend/run.hpp#L218 for an example of an accurate bounds checking.

I have a naive question about this: if this code is a better approach to check bounds than using `get_lower_bounds`/`get_upper_bounds`, wouldn't it make sense to extract this into a generic function so it can be used in different places?